### PR TITLE
Bug Fix: STENCIL-4087 - Preserve existing indexes 

### PR DESCRIPTION
**Pull request by tabayomi**
Regression was introduced in #71  that reindexes the querystring when it is being normalized to remove empty array values.

This PR addresses that bug.

@bigcommerce/stencil-team

### CODE REVIEW

**Comment by junedkazi** [comment]
@ransomcarroll @steve-ross can you try the fix in this PR to see if it resolves the issue.

**Comment by steve-ross** [comment]
@junedkazi works for me.
input: `action=add&attribute%5B121%5D=105&product_id=114&qty%5B%5D=1`
output `action=add&product_id=114&attribute%5B121%5D=105&qty%5B%5D=1`

**Comment by steve-ross** [comment]
I was too late to the game 😀 
https://github.com/steve-ross/stencil-utils/pull/1/commits/e753788c5206f489c9446ad5ebdec4da9c73498a

**Comment by steve-ross** [comment]
The test spec should probably be updated to compare serialized strings instead of deserialized. Might have caught this bug in that case.

File: src/lib/utils.js [hunk 1]
```diff
@@ -7,21 +7,25 @@ const queryString = require('query-string');
...
     const optionsCollection = Object.keys(qsParams).reduce((normalized, key) => {
         let param;
 
         if (qsParams[key] instanceof Array) {
-            const filtered = qsParams[key].filter((item) => (item !== '' && item !== undefined));
+            const filtered = qsParams[key].filter(isValidParam);
             param = { [key]: filtered };
         } else {
-            param = { [key]: qsParams[key] };
+            if (isValidParam(qsParams[key])) {
```
Comment (mjschock) [inline src/lib/utils.js]: nit; just pull the `if` into an `if else`

**Comment by creativesafetysupply** [comment]
Nothing changes still when changing product options (e.g. image, SKU, price), which is I'm guessing the intended outcome of this fix. bigcommerce/cornerstone#1109

**Comment by sacr3dc0w** [comment]
Nothing changes still when changing product options (e.g. image, SKU, price), which is I'm guessing the intended outcome of this fix. bigcommerce/cornerstone#1109

**Comment by tabayomi** [comment]
@siliconforks - Thanks for the feedback. You'll notice from the link below - the `params` attribute in the object has been removed since the `1.0.0` release - March last year.

https://github.com/bigcommerce/stencil-utils/blob/1.0.4/src/api/product-attributes.js#L24

**Comment by tabayomi** [comment]
@sacr3dc0w - Were you able to test using the fix implemented in this PR? If it's still not working, would you mind pasting a sample of your querystring in this thread so I can take a look?

**Comment by siliconforks** [comment]
I think maybe the syntax sugar in ECMAScript 6 is causing some confusion - changing `{ params: params }` to `{ params }` works because those are equivalent.  But `{ normalizedQs }` is equivalent to `{ normalizedQs: normalizedQs }` which is not what the `remoteRequest` method is expecting (it expects the key in the object to be named `params`, not `normalizedQs`).

**Comment by tabayomi** [comment]
@siliconforks - Thanks for clarifying that, you are quite right. The PR has been updated to reflect the attribute name. 

@sacr3dc0w - Please test with this new update and let me know if this fixes things.

File: src/test-unit/product.options.spec.js [hunk 2]
```diff
@@ -4,23 +4,22 @@ const queryString = require('query-string');
...
     });
 
-    it('should remove empty array members', () => {
-        const qs = 'attribute[0]=123&attribute[1]=&qty=1';
-        const normalized = normalizeQueryStringParams(qs);
-        const parsed = queryString.parse(normalized, { arrayFormat: 'index' });
-       
-        expect(parsed['attribute'].length).toBe(1);
-    });
+    it('should preserve array indexes on a valid querystring', () => {
```
Comment (ransomcarroll) [inline src/test-unit/product.options.spec.js]: great job on this test guys. o7

**Pull request merged**